vector/src/main/java/im/vector/app/features/userdirectory/UserListViewModel.kt

```diff
@@ -0,0 +1,262 @@
+ * limitations under the License.
+ */
+
+package im.vector.app.features.userdirectory
+
+import androidx.fragment.app.FragmentActivity
+import androidx.lifecycle.viewModelScope
+import com.airbnb.mvrx.ActivityViewModelContext
+import com.airbnb.mvrx.FragmentViewModelContext
+import com.airbnb.mvrx.Loading
+import com.airbnb.mvrx.MvRxViewModelFactory
+import com.airbnb.mvrx.Success
+import com.airbnb.mvrx.ViewModelContext
+import com.jakewharton.rxrelay2.BehaviorRelay
+import com.squareup.inject.assisted.Assisted
+import com.squareup.inject.assisted.AssistedInject
+import im.vector.app.core.contacts.ContactsDataSource
+import im.vector.app.core.contacts.MappedContact
+import im.vector.app.core.extensions.exhaustive
+import im.vector.app.core.extensions.toggle
+import im.vector.app.core.platform.VectorViewModel
+import im.vector.app.features.createdirect.CreateDirectRoomActivity
+import im.vector.app.features.invite.InviteUsersToRoomActivity
+import io.reactivex.Single
+import io.reactivex.android.schedulers.AndroidSchedulers
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.launch
+import org.matrix.android.sdk.api.MatrixCallback
+import org.matrix.android.sdk.api.session.Session
+import org.matrix.android.sdk.api.session.identity.FoundThreePid
+import org.matrix.android.sdk.api.session.identity.ThreePid
+import org.matrix.android.sdk.api.util.toMatrixItem
+import org.matrix.android.sdk.rx.rx
+import timber.log.Timber
+import java.util.concurrent.TimeUnit
+
+private typealias KnownUsersSearch = String
+private typealias DirectoryUsersSearch = String
+private typealias ContactUsersSearch = String
+
+class UserListViewModel @AssistedInject constructor(@Assisted
+                                                    initialState: UserListViewState,
+                                                    private val contactsDataSource: ContactsDataSource,
+                                                    private val session: Session)
+    : VectorViewModel<UserListViewState, UserListAction, UserListViewEvents>(initialState) {
+
+    private val knownUsersSearch = BehaviorRelay.create<KnownUsersSearch>()
+    private val directoryUsersSearch = BehaviorRelay.create<DirectoryUsersSearch>()
+    private val contactUsersSearch = BehaviorRelay.create<ContactUsersSearch>()
+
+    private var allContacts: List<MappedContact> = emptyList()
+    private var mappedContacts: List<MappedContact> = emptyList()
+
+    @AssistedInject.Factory
+    interface Factory {
+        fun create(initialState: UserListViewState): UserListViewModel
+    }
+
+    companion object : MvRxViewModelFactory<UserListViewModel, UserListViewState> {
+        override fun create(viewModelContext: ViewModelContext, state: UserListViewState): UserListViewModel? {
+            return when (viewModelContext) {
+                is ActivityViewModelContext -> {
+                    when (viewModelContext.activity<FragmentActivity>()) {
```

Comment:
use when (val activity = viewModelContext.act...) in order to have smart cast?

Reply:
Done, thanks.